proposals/intraprocess_communication.md

```diff
@@ -0,0 +1,525 @@
+
+![Current IPC Block Diagram](old_ipc.png)
+
+Several shortcomings of the current implementation are listed below.
+
+### Incomplete Quality of Service support
+
+The current implementation can't be used when the QoS durability value is set to `Transient Local`.
+
+The current implementation does not enforce the depth of the QoS history in a correct way. The reason is that
+there is a single ring buffer per `Publisher` and its size is equal to the depth of the `Publisher`'s history.
+A `Publisher` stores a message in the ring buffer and then it sends a meta-message to allow a `Subscription` to retrieve it.
+The `Subscription` correctly stores meta-messages up to the number indicated by its depth of the history, but,
+depending on the frequency at which messages are published and callbacks are triggered, it may happen that a
+meta-message processed from the `Subscription` does not correspond anymore to a valid message in the ring
+buffer, because it has been already overwritten.
+
+More details [here](https://index.ros.org/doc/ros2/Concepts/About-Quality-of-Service-Settings/).
+
+**TODO:** take into account also new QoS: Deadline, Liveliness and Lifespan
+[reference](https://github.com/ros2/design/pull/212).
+
+### Dependent on the RMW
+
+The current implementation has to send meta-messages from the `Publisher` to the `Subscription`s. This is done
+using the `rmw_publish` function, the implementation of which depends on the chosen middleware. This results
+in having the performance of a ROS2 application with intra-process communication enabled to still heavily
+depend on the chosen RMW implementation.
+
+At the moment, this results in a relevant impact on the performance of the application for
+all the ROS2 supported middlewares (Fast-RTPS, OpenSplice, Connext, DPS).
+
+##### Memory requirement
+
+When a `Node` creates a `Publisher` or a `Subscription` to a topic `/MyTopic`, it will also create an
+additional one to the topic `/MyTopic/_intra`. The second topic is the one where meta-messages travel. Our
+[experimental results](https://github.com/irobot-ros/ros2-performance/tree/master/performances/experiments/crystal/pub_sub_memory#adding-more-nodes-x86_64)
+show that creating a `Publisher` or a `Subscription` has a non-negligible memory cost.
+
+Moreover, when using the default RMW implementation, Fast-RTPS, the memory consumed by a ROS2 application
+increases almost exponentially with the number of `Publisher` and `Subscriptions`.
+
+##### Latency and CPU utilization
+
+Publishing a meta-message has the same overhead as that of publishing a small inter-process message.
+
+However, comparing the publication/reception of an intra and an inter-process message, the former requires
+several additional operations: it has to store the message in the ring buffer, monitor the number of
+`Subscription`s, and extract the message.
+
+The result is that from the latency and CPU utilization point of view, it is convenient to use inter-process
```

Comment:
Do you mean `intra-process` here, or am I misunderstanding?

Reply:
Yes, I meant `intra-process`